### PR TITLE
ENH: `differentiate`: add sub-package for array-API compatible numerical differentiation tools

### LINKED ISSUE #2035: Add methods to compute derivatives to Scipy (Trac #1510)
_Original ticket http://projects.scipy.org/scipy/ticket/1510 on 2011-09-05 by trac user deil, assigned to unknown._

It would be nice if scipy included numerical differentiation methods.
I'm not sure where it would fit best, maybe in scipy.optimize or in scipy.misc?

scipy.optimize already contains an approx_fprime method, which approximates
the Jacobian. Having a second method approx_hessian to compute the Jacobian
would be very useful, because the covariance matrix can be computed as the
inverse of the Hessian, and having this in scipy would make it possible to compute
fit parameter errors for other optimizers than scipy.optimize.leastsq.

There are already two numpy-based python packages that implement various variants of  numerical differentiation routines:

http://code.google.com/p/numdifftools/

https://github.com/statsmodels/statsmodels/blob/master/scikits/statsmodels/sandbox/regression/numdiff.py

I am not enough familiar with the differences in the methods implemented to make a suggestion which methods would be most useful to add to scipy.
Maybe the authors can comment?

The need for Hessians came up in this mailing list discussion:

http://mail.scipy.org/pipermail/scipy-user/2011-September/030459.html

### LINKED ISSUE #17059: ENH: Robust and fast numerical derivative for error propagation
### Is your feature request related to a problem? Please describe.

I am a particle physicist by trade, and error propagation is very important for us. We need to propagate errors through complex functions that may execute arbitrary code and whose analytical behavior we do not understand in detail. Therefore, it is important to have a reliable and fast black-box derivative that just works and does not require expert knowledge about the function.

### Describe the solution you'd like.

I wrote the [jacobi library](https://github.com/hDembinski/jacobi) ([docs](https://hdembinski.github.io/jacobi/)) which I would like to include in scipy. Jacobi is an implementation of John d'Errico's DERIVEST algorithm for matlab written from scratch. Jacobi has remarkable features that make it ideal for this use case. It is a fast, safe-to-use, black-box numerical derivative for many applications. I think it is a great match for a foundational scientific library like scipy. On top of the derivative, I wrote a function `propagate` which uses the derivative to do error propagation for any vector-valued function.

Jacobi uses an adaptive algorithm to compute the numerical derivative. It converges fast if the derivative is trivial (constant or linear) and it can handle functions which have poor numerical accuracy (they are numerically noisy, use single precision arithmetic etc.). The amount of numerical noise is detected automatically and propagated into an error estimate for the derivative, which is computed alongside the derivative. This is fairly unique for such a library. By default, the algorithm runs until the error on the result does not improve anymore (usually when it reaches machine precision), but one can also set an accuracy target and jacobi will stop once the estimated error drops below this target.

jacobi uses polynomial extrapolation, which is a higher-order form of Richardson extrapolation. Thanks to extrapolation, it can compute numerical derivatives to machine precision. The polynomial fit also provides the error estimate on the extrapolated derivative.

jacobi is written in pure Python and the implementation is written in a concise way by me, a Python expert and author and maintainer of several open source Python and C++ libraries. I am also the designer and author of Boost Histogram in the Boost C++ project. Jacobi uses only 250 lines. It is very fast, too, because it uses numpy array math efficiently. It is faster than numdifftools by up to a factor of 2 to 1000, depending on the problem.

### Describe alternatives you've considered.

- **scipy.misc.derivative** is too difficult to use correctly. The step size `dx` has to be provided by the user, it is not determined automatically. I cannot run this function like a black box and get my derivative to machine precision, I need to know what I am doing and set the parameters `dx` and `order` to suitable values, which I can only do optimally if I already know a lot about my function and its implementation. jacobi is a black-box derivative. It just works, parameters like step size and order are determined automatically from an iterative analysis of the function response.
- **scipy.optimize.approx_fprime** or **scipy.optimize.approx_derivative** are not robust, cannot achieve machine precision unless "cs" step is used which is rarely possible in practice, and do not provide an error estimate. They call a low-order finite difference formula, which has sqrt(EPS) precision at best (unless "cs" step is used, which is rarely possible in practice), while jacobi often achieves EPS. They are not robust, because these algorithms cannot detect that the function is noisy; the step size that each algorithm selects to find the optimum between truncation error and round-off error is only valid for a function that calculates the correct result within a few ULPs. Non-trivial functions often do not achieve this. It is also possible that some part of the internal calculation is done in single precision, but the inputs and outputs are double precision. In that case, neither algorithm can detect the right step size. jacobi handles this case just fine, because it detects the precision of the calculation by probing the function response iteratively.
- **numdifftools** offers a black-box derivative, but it is neither robust nor fast. It does not provide an error estimate. The implementation is also much more complex than jacobi due to excessive use of code structuring.
- **automatic differentiation** libraries like JAX are very intrusive, this is bad. They cannot be integrated into an existing code base, one has to write the application around the AD library. Jacobi can be integrated into any application and it can calculate the derivative for any function, whatever the implementation of that function does (it could call into C or Fortran, use numba, etc.).

### Additional context (e.g. screenshots, GIFs)

Jacobi is not intended to provide first derivatives for numerical minimizers like those accessible via scipy.optimize.minimize. It is not a replacement for **scipy.optimize.approx_fprime** or **scipy.optimize.approx_derivative**. Numerical minimizers are designed to handle imperfect derivatives that can be computed very fast. Jacobi is not fast enough for those applications and too accurate.

Check out https://hdembinski.github.io/jacobi/ for plots and benchmarks that show how well jacobi works.

### PR DESCRIPTION
**Pull request by mdhaber**
#### Reference issue
Closes gh-2035
Closes gh-17059

#### What does this implement/fix?
This PR moves the private `_differentiate` and `_jacobian` and associated tests from `scipy.optimize._differentiate` to a new, public subpackage. There are already other tools for numerical differentiation in the ecosystem; nonetheless, there has been a lot of support for adding such functions to SciPy over the years. Also, IIUC, the existing tools may not be vectorized or array API compatible like these are.  More thorough discussion about inclusion of such functions in SciPy is in gh-17059.

#### Additional information
The `differentiate` function already supports CuPy and PyTorch arrays. `jacobian` will be upgraded separately after this merges. `hessian` is under review in gh-20708.

I know there are concerns about absolute vs relative steps. There are a few other algorithmic improvements that we might like to make before the 1.15 release, but I'd suggest that we open a separate issue about these sorts of things once this merges.

### CODE REVIEW

**Comment by lucascolley** [comment]
I agree with the naming of `scipy.differentiate` 👍

File: scipy/differentiate/__init__.py
Comment (lucascolley) [inline scipy/differentiate/__init__.py:3]:
```suggestion
Finite Difference Differentiation (:mod:`scipy.differentiate`)
```

File: scipy/differentiate/_differentiate.py
Comment (lucascolley) [inline scipy/differentiate/_differentiate.py:77]:
```suggestion
         with `x`. `f` must be an elementwise function: each element
```

File: scipy/differentiate/_differentiate.py
Comment (lucascolley) [inline scipy/differentiate/_differentiate.py:79]:
```suggestion
        the array `x` or the arrays in ``fargs``.
```

**Comment by rgommers** [comment]
Thanks @mdhaber.
 
> I agree with the naming of `scipy.differentiate` 👍

+1 I think that's a good name, and beats the alternatives I can remember from previous discussions.

* * *

The diff here looks nice and concise, and it seems enough people want this functionality made public. So it seems like a win to me. I'm not familiar with the implementation, so I'll rely on the judgement of domain experts here. Adding the new submodule seems fine to me - I trust that we can keep it as high-quality and well-tested code given the maintainer interest I see.

File: scipy/differentiate/_differentiate.py
Comment (mdhaber) [reply scipy/differentiate/_differentiate.py:77]: Haha yes but also
```suggestion
         with ``x``. `f` must be an elementwise function: each element
```
Because here `x` refers to the argument passed to `f`, which is not the same as the argument passed to `differentiate`. I'll give it a different name.

File: scipy/differentiate/_differentiate.py
Comment (mdhaber) [reply scipy/differentiate/_differentiate.py:79]: Actually I think this one is right, but I need to change the name. https://github.com/scipy/scipy/pull/21041/files#r1652926001

File: scipy/differentiate/_differentiate.py
Comment (mdhaber) [inline scipy/differentiate/_differentiate.py:814]:
```suggestion
    >>> from scipy.differentiate import jacobian
```

**Review by steppi — APPROVED** [review]
This looks good to me. I had reviewed the algorithms carefully for https://github.com/scipy/scipy/pull/20630, https://github.com/scipy/scipy/pull/20063, and https://github.com/scipy/scipy/pull/18811, and think the approach is fundamentally sound and seems to work well in practice.

I have some thoughts about using absolute vs relative steps but think it's fine to leave that for the future following @mdhaber's suggestion.

I had a question here, https://github.com/scipy/scipy/pull/20834#discussion_r1667737498,  about documenting parameters as taking `array` instead of `array_like`, but still accepting lists and other array like things. Is there a reference or SPEC or something that says this is what we're supposed to do? I don't consider this a blocker though.

**Comment by lucascolley** [comment]
> I had a question here, https://github.com/scipy/scipy/pull/20834#discussion_r1667737498, about documenting parameters as taking array instead of array_like, but still accepting lists and other array like things.

My take is that we should say `array` for new APIs, but not explicitly put in effort to reject array-likes. That would mean that we officially support only arrays as inputs and don't give BC guarantees about array-likes; but we may as well continue to unofficially support array-likes where it is easy/natural.

That gives us flexibility to go either way in the future on strictly accepting only arrays or officially supporting array-likes. I do think we would need to be careful to communicate this well whenever any BC-breaks actually occur (even if the BC was never guaranteed by the type in the docstring), but that should be doable.

Does that sound reasonable @rgommers @mdhaber @steppi ? *If* we decided to move towards array-only in (say) SciPy 2.0, the first step could then be to change `array_like` to `array` in existing docstrings.

**Comment by rgommers** [comment]
That doesn't sound right to me, it just sets us up for a problem later on. We should do one of two things:
1. accept array-likes after all, or
2. ensure we validate that the input is an array, and give a good error message otherwise (e.g. check that the input has a `.dtype` attribute, and raise an exception otherwise).

**Comment by mdhaber** [comment]
> just sets us up for a problem later on.

To make sure we're on the same page, which problem(s)?

**Comment by rgommers** [comment]
> To make sure we're on the same page, which problem(s) did you have in mind?

Lucas wrote _"not explicitly put in effort to reject array-likes ... and don't give BC guarantees about array-likes"_. If we do that, list inputs may work at first then stop working. That's just never a good idea - regressions downstream and bug reports are bound to follow.




> Also, can we clarify what we mean by `array_like`? The formal definition, whether one looks in the NumPy glossary or `numpy.typing`, is only a little more specific than `object`. Does `real array_like`, meaning an object that is coerced to an array with real dtype, work?

There's only an implicit definition: if it gets turned into an ndarray (of a useful dtype) by calling `np.asarray` on it, then it's an array-like. This could be for example some custom class instance with a `__array__` method - whatever works really.

**Comment by mdhaber** [comment]
> If we do that, list inputs may work at first then stop working.

So another option would be to emit a deprecation warning if and when the behavior is going to change. There would be a few other functions doing that, right?

> There's only an implicit definition:

That agrees with the definition in the NumPy glossary and `numpy.typing`, but it is quite broad, so I would prefer to be more specific if we were to change this to something like option 1. How about "array_like of floats" rather than just "array_like"?

**Comment by steppi** [comment]
@rgommers, do you think it's OK that I merged this before `array` vs `array_like` was settled? I saw your positive comment [here](https://github.com/scipy/scipy/pull/21041#issuecomment-2188583744) and thought it would be fine to merge since I'd reviewed the algorithm closely. Do you think I should hold off on merging this other one, https://github.com/scipy/scipy/pull/20834, until the `array` matter is settled? I'm not well informed on this debate, and have been reviewing only from an algorithmic perspective.

**Comment by rgommers** [comment]
> do you think it's OK that I merged this before array vs array_like was settled? 

It's not a real problem, so don't worry about it too much. Figuring it out before merging would have been nicer though, since it is an API design question that was raised before merging, so next time I'd say hold off on hitting the green button.

>  How about "array_like of floats" rather than just "array_like"?

That sounds fine to me. It's more or less the same; we don't really care if the "real numbers" or "of floats" is in the type specifier or in the description right below it - either way works. The important thing is the `array` vs. `array_like`.

I now had an actual look at the code in question, for `differentiate`. The validation for `x` is coming from `array_namespace`, which accepts array-likes, and this is tested (at least for Python scalar input, e.g. with `@pytest.mark.parametrize('x', [0.6, np.linspace(-0.05, 1.05, 10)])`).

So, a small tweak to change the docs from "`array`" to "`array_like`" will do it.

Changing the code to match `array` instead would be a larger change. It's very rare that we add a new submodule, so this is an opportunity to do it. But I personally don't have much time to invest in this, so if there are no other takers then it's also fine to not investigate this more.

**Comment by steppi** [comment]
> Changing the code to match `array` instead would be a larger change. It's very rare that we add a new submodule, so this is an opportunity to do it. But I personally don't have much time to invest in this, so if there are no other takers then it's also fine to not investigate this more.

I'm out on the loop on this and am not really aware of the motivation for wanting to be more strict and only allow array inputs. Do you consider that to be the eventual desired behavior for everything?  Are there some things I could read to get up to speed? If this is a goal for SciPy 2.0, maybe there's even actually a benefit to doing it in lockstep rather than having a temporary period where we maintain a mix of the old behavior and the new behavior -- at least for new users who might be tripped up by having some things being more strict than others, and may not have the context that it relates to when things were added. I guess punting this all to SciPy 2.0 is only feasible though if there aren't other important changes we'd like to make that would need an even longer timescale to complete. If the only option is to do things piecemeal a few deprecations at a time, making a new submodule like this have the more strict behavior seems like the way to go, if that's the eventual desired state.

**Comment by rgommers** [comment]
> Do you consider that to be the eventual desired behavior for everything?

If I'd write new code for some project from scratch, definitely. It solves a lot of problems, from making ndarray subclasses more useful to sane type annotations. It's also telling that no other array/tensor libraries emulated numpy's anything-goes array-likes (they all require instances of their own array/tensor classes).

That said, I am not sure it's something we want to do for all of SciPy, because it'd be a _lot_ of churn. It could be a per-submodule thing as well perhaps, which is why this new submodule could be an opportunity.

**Comment by lucascolley** [comment]
I think that, for example, having `integrate` accept array-likes but `differentiate` reject them would be quite confusing for users. I would definitely prefer if we could do it project-wide (which shouldn't be too difficult once everything is using `array_namespace`). It would be a huge amount of churn though, so feels like a SciPy 2.0 thing to me.

**Comment by steppi** [comment]
In https://github.com/scipy/scipy/pull/20834 for `nsum`, @mdhaber is updating the documentation from `array` to (appropriately types) `array_like` in both `nsum` and `differentiate`, which I think is an acceptable resolution.

**Pull request merged**
